### ISSUE TITLE
Validation rules for internal sets

### ISSUE THREAD

**Issue by ghost** [makumba/makumba#563]
**Reported by @rudolfmayer on 10 Dec 2007 18:10 UTC**
For internal sets, a syntax needs to be defined for how to define validation rules, and how to refer to the parent object. 

For example:

Given a data definition best.internal.Server with 

accounts=set
accounts->person=ptr general.Person
accounts-> ....

i would want to define the rule that for each server, each person can be added only once, thus we need to define a multi-field key.

validation rules in a set could be defined simply as normal:
accounts->uniqueAccount%unique = parent, person : An account for this person on this server exists already!

The syntax should even pass the mdd parser already, just the keyword to access the parent object should be defined - i proposed here parent (this would imho refer to the set entry itself), but that is still discussable of course

Migrated-From: http://trac.makumba.org/ticket/1017

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 20 Jul 2008 10:40 UTC**
functions already work this way
subfield->sessionVar%function(params){expression}error;comment

should be easy to enhance the record parser for validation to be parsed like functions are

**Comment by ghost** [comment]
**Comment by @rudolfmayer on 29 Aug 2008 23:32 UTC**
(In reply to comment #1)
> functions already work this way
> subfield->sessionVar%function(params){expression}error;comment
> 
> should be easy to enhance the record parser for validation to be parsed like
> functions are


record parser now works with validation definitions in subsets, still the initial issue from comment 0 still exists - how to refer to the parent object?

**Comment by ghost** [comment]
**Comment by @manuelbernhardt on 23 Feb 2010 11:41 UTC**
this issue also exists for QFs:

events= set 
events->who= fixed not null ptr projman.Principal
events->type= int{"create"=0, "modify"=1, "view"=2, "markAsView"=3, "remove"=10, "aprobatPG"=31, "aprobatExe"=32, "aprobatBeneficiar"=33, "marcatExecutat"=34, "nuSePoateRemedia"=35}
events->comment= text

events->didPrincipalKnow(){ who=actor(projman.Principal) AND type IN (0, 2,3) }


i.e. "who" and "type" are not part of the super-type but of the set.

hence the correct syntax for the QF would IMO be

events->didPrincipalKnow(){ this.who=actor(projman.Principal) AND this.type IN (0, 2,3) }

as for refering to the parent type, is there maybe a semantic from other compilers (e.g. the java compiler) that supports that case? "super" is for inheritance so I would not go for it, I mean we are talking about a nested member.

note that if we use "parent" that means "parent" becomes a reserved keyword, which is fine by me.

**Comment by ghost** [comment]
**Comment by @rudolfmayer on 23 Feb 2010 11:44 UTC**
(In reply to comment #3)
> events= set 
> events->who= fixed not null ptr projman.Principal
> events->type= int{"create"=0, "modify"=1, "view"=2, "markAsView"=3,
> "remove"=10, "aprobatPG"=31, "aprobatExe"=32, "aprobatBeneficiar"=33,
> "marcatExecutat"=34, "nuSePoateRemedia"=35}
> events->comment= text
> 
> events->didPrincipalKnow(){ who=actor(projman.Principal) AND type IN (0, 2,3) }
> 
> 
> i.e. "who" and "type" are not part of the super-type but of the set.
> 
> hence the correct syntax for the QF would IMO be
> 
> events->didPrincipalKnow(){ this.who=actor(projman.Principal) AND this.type IN
> (0, 2,3) }
> 
> as for refering to the parent type, is there maybe a semantic from other
> compilers (e.g. the java compiler) that supports that case? "super" is for
> inheritance so I would not go for it, I mean we are talking about a nested
> member.
> 
> note that if we use "parent" that means "parent" becomes a reserved keyword,
> which is fine by me.


in Java, when in a nested class you want to refer ecplicitely to something from the enclosing class, you normally would use

EnclosingClassName.this.fieldName

**Comment by ghost** [comment]
**Modified by @manuelbernhardt on 28 Jun 2010 07:22 UTC**

**Comment by ghost** [comment]
**Modified by @manuelbernhardt on 28 Jun 2010 07:36 UTC**

**Comment by ghost** [comment]
**Modified by @manuelbernhardt on 28 Jun 2010 14:19 UTC**

**Comment by ghost** [comment]
**Comment by @rudolfmayer on 14 Dec 2010 22:45 UTC**
Some design discussion equivalent to this is also in #513

**Comment by ghost** [comment]
**Comment by @rudolfmayer on 15 Dec 2010 13:11 UTC**
An observation from implementation side:

using the MDD name of the enclosing type, e.g in test.Person

address->unique(description, Person) : "Address descriptions must be unique per person!"

is very convenient for not-unique definitions - parsing is relatively easy, and as the subset already contains the automatic field "Person", creating the unique indices on the DB table works right away.

For unique definitions that would actually make use of a field outside the subset, and thus cannot be defined on the database, but require a query, most likely more processing is needed. but these seem to be the more rare cases, and thus this could be moved to another bug entry.
